src/util/file-filter.js

```diff
@@ -0,0 +1,110 @@
+/* @flow */
+import path from 'path';
+
+import minimatch from 'minimatch';
+
+import {createLogger} from './logger';
+
+const log = createLogger(__filename);
+
+// Use this function to mimic path.resolve without resolving to absolute path.
+export const normalizeResolve = (file: string): string => {
+  // normalize
+  file = path.normalize(file);
+
+  // trim trailing slash
+  if (path.parse(file).base && file.endsWith(path.sep)) {
+    return file.slice(0, -1);
+  }
+  return file;
+};
+
+// FileFilter types and implementation.
+
+export type FileFilterOptions = {|
+  filesToIgnore?: Array<string>,
+  ignoreFiles?: Array<string>,
+  sourceDir?: string,
+  artifactsDir?: string,
+|};
+
+/*
+ * Allows or ignores files.
+ */
+export class FileFilter {
+  filesToIgnore: Array<string>;
+  sourceDir: string;
+
+  constructor({
+    filesToIgnore = [
+      '**/*.xpi',
+      '**/*.zip',
+      '**/.*', // any hidden file and folder
+      '**/.*/**/*', // and the content inside hidden folder
+      '**/node_modules',
+      '**/node_modules/**/*',
+    ],
+    ignoreFiles = [],
+    sourceDir = '',
+    artifactsDir,
+  }: FileFilterOptions = {}) {
+
+    this.filesToIgnore = [];
+    this.sourceDir = sourceDir;
+
+    this.addToIgnoreList(filesToIgnore);
+    if (ignoreFiles) {
+      this.addToIgnoreList(ignoreFiles);
+    }
+    if (artifactsDir) {
+      this.addToIgnoreList([
+        artifactsDir,
+        path.join(artifactsDir, '**/*'),
+      ]);
+    }
+  }
+
+  /**
+   *  Resolve relative path to absolute path if sourceDir is setted.
+   */
+  resolve(file: string): string {
+    if (this.sourceDir) {
+      return path.resolve(this.sourceDir, file);
+    }
+    return normalizeResolve(file);
+  }
+
+  /**
+   *  Insert more files into filesToIgnore array.
+   */
+  addToIgnoreList(files: Array<string>) {
+    for (const file of files) {
+      this.filesToIgnore.push(this.resolve(file));
+    }
+  }
+
+  /*
+   * Returns true if the file is wanted.
```

Comment:
I think we should add another sentence to this comment (if you agree that it is correct):

If `path` does not start with a slash, it will be treated as a path relative to `sourceDir` when matching it against all configured ignore-patterns.